core/utils.go

```diff
@@ -283,69 +88,42 @@ func createUser(fn string) {
 	}
 }
 
-func DownloadFile(filepath string, url string, spinner *pterm.SpinnerPrinter) error {
-	response, err := http.Head(url)
-	if err != nil {
-		return err
-	}
-	size, _ := strconv.Atoi(response.Header.Get("Content-Length"))
-
-	resp, err := http.Get(url)
-	if err != nil {
-		return err
-	}
-
-	defer func(Body io.ReadCloser) {
-		err := Body.Close()
-		if err != nil {
-			log.Fatalf("Error while closing HTTP stream: %v", err)
-		}
-	}(resp.Body)
-
-	out, err := os.Create(filepath)
-	if err != nil {
-		return err
-	}
-
-	defer func(out *os.File) {
-		err := out.Close()
-		if err != nil {
-			log.Fatalf("Error while closing output file: %v", err)
-		}
-	}(out)
-
-	buffer := make([]byte, 1024)
-	total := 0
-	lastTotal := 0
-	text := ""
-	if spinner != nil {
-		text = spinner.Text
-	}
-	for {
-		length, err := resp.Body.Read(buffer)
-		if err != nil && err != io.EOF {
-			return err
-		}
-		total += length
-		if spinner != nil && total-lastTotal > 1024*1024 {
-			lastTotal = total
-			spinner.UpdateText(fmt.Sprintf("%s (%d %%)", text, 100*total/size))
-		}
-		if length == 0 {
-			break
-		}
-		if _, err = out.Write(buffer[:length]); err != nil {
-			return err
+func writeFileIfNew(filepath string, content string) {
+	if _, err := os.Stat(filepath); os.IsNotExist(err) {
+		if err := os.WriteFile(filepath, []byte(content), 0o755); err != nil {
+			log.Fatal(err)
 		}
 	}
+}
 
-	if total != size {
-		return fmt.Errorf("downloaded file size doesn't match expected size")
+func getPluginIds(plugins []platform.Plugin) []string {
+	ids := make([]string, len(plugins))
+	for i, plugin := range plugins {
+		ids[i] = plugin.Id
 	}
+	return ids
+}
 
-	if spinner != nil {
-		spinner.UpdateText(fmt.Sprintf("%s (100 %%)", text))
+func (o *QodanaOptions) guessProduct() string {
+	if o.Ide != "" {
+		productCode := strings.TrimSuffix(o.Ide, EapSuffix)
+		if _, ok := Products[productCode]; ok {
+			return productCode
+		}
+		return ""
+	} else if o.Linter != "" {
+		// if Linter contains registry.jetbrains.team/p/sa/containers/ or https://registry.jetbrains.team/p/sa/containers/
```

Comment:
I think this would fail for custom registries? Companies often use those to save money on docker pulls (OTOH, no one complained so far, and I guess it's out of scope of this PR)